src/polygon/src/polygon.cpp

```diff
@@ -192,900 +165,486 @@ void PrintFailure(const std::exception &e, const PolygonsIdx &polys,
-     * polygon is not monotone, as the result should be topologically valid, but
-     * not geometrically. The parameter, last, must be set true only for the
-     * final point, as this ensures the last triangle is output.
-     */
-    void ProcessVert(const VertItr vi, bool onRight, bool last,
-                     std::vector<glm::ivec3> &triangles) {
-      VertItr v_top = reflex_chain_.back();
-      if (reflex_chain_.size() < 2) {
-        reflex_chain_.push_back(vi);
-        onRight_ = onRight;
-        return;
+
+    float Cost(VertItr v, glm::vec2 openSide, float precision) const {
+      const glm::vec2 offset = v->pos - pos;
+      float cost = SignedDist(v, rightDir, precision);
+      if (glm::isfinite(cost)) {
+        cost = glm::min(cost, SignedDist(v, left->rightDir, precision));
       }
-      reflex_chain_.pop_back();
-      VertItr vj = reflex_chain_.back();
-      if (onRight_ == onRight && !last) {
-        // This only creates enough triangles to ensure the reflex chain is
-        // still reflex.
-        PRINT("same chain");
-        int ccw = CCW(vi->pos, vj->pos, v_top->pos, precision_);
-        while (ccw == (onRight_ ? 1 : -1) || ccw == 0) {
-          AddTriangle(triangles, vi, vj, v_top);
-          v_top = vj;
-          reflex_chain_.pop_back();
-          if (reflex_chain_.empty()) break;
-          vj = reflex_chain_.back();
-          ccw = CCW(vi->pos, vj->pos, v_top->pos, precision_);
+      if (glm::isfinite(cost)) {
+        float openCost =
+            glm::determinant(glm::mat2(openSide, v->pos - right->pos));
+        if (cost == 0 && glm::abs(openCost) < precision) {
+          return kBest;
         }
-        reflex_chain_.push_back(v_top);
-        reflex_chain_.push_back(vi);
-      } else {
-        // This branch empties the reflex chain and switches sides. It must be
-        // used for the last point, as it will output all the triangles
-        // regardless of geometry.
-        PRINT("different chain");
-        onRight_ = !onRight_;
-        VertItr v_last = v_top;
-        while (!reflex_chain_.empty()) {
-          vj = reflex_chain_.back();
-          AddTriangle(triangles, vi, v_last, vj);
-          v_last = vj;
-          reflex_chain_.pop_back();
+        cost = glm::min(cost, openCost);
+      }
+      return cost;
+    }
+
+    float DelaunayCost(glm::vec2 diff, float scale, float precision) const {
+      return -precision - scale * glm::dot(diff, diff);
+    }
+
+    float EarCost(float precision) const {
+      glm::vec2 openSide = left->pos - right->pos;
+      const glm::vec2 center = 0.5f * (left->pos + right->pos);
+      const float scale = 4 / glm::dot(openSide, openSide);
+      openSide = glm::normalize(openSide);
+
+      float totalCost = glm::dot(left->rightDir, rightDir) - 1 - precision;
+      if (CCW(pos, left->pos, right->pos, precision) == 0) {
+        return totalCost < -1 ? kBest : 0;
+      }
+      VertItr test = right;
+      while (test != left) {
```

Comment:
This loop is what makes ear clipping O(n^2). I haven't been paying very close attention to the change in performance; hopefully it's not too bad, though I bet a few of our test cases will probably show it more dramatically than the common cases.